olaaf_django/sync_hashes.py

```diff
@@ -87,22 +96,51 @@ def _insert_diff_hashes(repo, prev_commit, current_commit):
       # if file was added or modified, calculate the new hash
       file_type = 'html' if file_name.endswith('.html') else 'pdf'
       if file_type == 'html':
-        url = path.rsplit('.', 1)[0]
+        if 'index.html' in path:
+          url = path.split('index.html')[0]
+          if url:
+            url = url.rsplit('/', 1)[0]
+        else:
+          url = path.rsplit('.', 1)[0]
+        hash_type = Hash.RENDERED
+
       else:
         url = path
+        hash_type = Hash.BITSTREAM
       # if file aready existed and it was modified or deleted update previous hash
       if action != 'A':
-        previous_hash = Hash.objects.get(path=url, end_commit__isnull=True)
-        previous_hash.end_commit = current_commit
-        previous_hash.save()
+        q = Q(path=url, end_commit__isnull=True)
+        if query is None:
+          query = q
+        else:
+          query = query | q
+        end_commits_by_path[url] = current_commit
       if action != 'D':
-        hash_value = _calculate_file_hash(repo, path, current_commit, file_type)
+        hash_value, search_path = _calculate_file_hash_and_search_path(repo, path, current_commit, file_type)
         if hash_value is not None:
-          h = Hash(value=hash_value, path=url, start_commit=current_commit)
-          h.save()
+          h = Hash(value=hash_value, path=url, hash_type=hash_type)
+          new_hashes.append(h)
+
+  updated_hashes = []
+  if query is not None:
+    updated_hashes = Hash.objects.filter(query)[::1]
```

Comment:
I believe you can just do `list(Hash.objects.filter(query))` which I think is much easier to understand what you are trying to do.

Reply:
but the more I think about this the more I think this should be properly solved with generators
```
def update_hash(h):
    h.end_commit = end_commits_by_path[h.path]
    return h

hashes_to_update = Hash.objects.filter(query).iterator() # default batch size 2000
hashes_pending_update = (update_hash(h) for h in hashes_to_update)
Hash.objects.bulk_update(hashes_pending_update, ['end_commit'], batch_size=2000)
```